### ISSUE TITLE
RecurrentCritic example

### ISSUE THREAD

**Issue by jaak-s** [aai-institute/tianshou#793]
+ [X] documentation request (i.e. "X is missing from the documentation.")
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates

I tried to use PPO+LSTM and got RecurrentActorProb working with PPO.

However, I was not able to make `RecurrentCritic` work with PPO. There was no concrete example anywhere in the documentation. If someone has I would be interested in a full example with PPO with both recurrent critic and recurrent actor.

**Comment by qtomcatq** [comment]
facing the same problem... and I'm also a researcher, not a pro coder to hustle this package

**Comment by Trinkle23897** [comment]
Here is the background:
- Currently, tianshou keeps "state" in sampling phase, so you can get the "state" from `policy.forward` function
- However, it won't keep that state and stored in replay buffer (of course you can manually do it by `info["model_state"] = state.detach()` somewhere, but keep in mind that's only actor state, no critic state is calculated)
- For training phase, in `policy.learn` function, you cannot find any "state"-related information. So all Recurrent* training use 0 state initialization

For RecurrentCritic support, the current codebase has already checked the input dimension (https://github.com/thu-ml/tianshou/blob/774d3d8e833a1b1c1ed320e0971ab125161f4264/tianshou/utils/net/continuous.py#L345-L351) to detect whether is a stacked observation or single-frame observation automatically. So it should work OOTB. Can you post more information about the error you run into?

BTW I personally don't favor RecurrentCritic because:
- Critic network is not used in inference time, only the actor network needs to keep a state;
- For training, you can always use a network with `obs[t-n:t] -> value` by some mapping instead of using recurrent network to calculate `obs[t-n], 0 -> ... -> obs[t], state[t] -> value` since the initial state for `obs[t-n]` is always 0.

**Comment by jaak-s** [comment]
Thank you for the response. Indeed, I got it working now. I previously had passed a non-zero action dimension to `RecurrentCritic`, which created a mismatching matrix shapes for the critic output layer. When using `[0]` (the default) for the action space of the ReucrrentCritic all is good now.

> However, it won't keep that state and stored in replay buffer (of course you can manually do it by `info["model_state"] = state.detach()` somewhere, but keep in mind that's only actor state, no critic state is calculated)

Does that mean the recurrent critic does not use any recurrent information? Or only its starting point is 0 hidden vector and the critic still uses the `stack_num` amount of state updates?

> For training, you can always use a network with `obs[t-n:t] -> value` by some mapping instead of using recurrent network to calculate `obs[t-n], 0 -> ... -> obs[t], state[t] -> value` since the initial state for `obs[t-n]` is always 0.

If I understand the proposal correctly this would mean that at each step t we compute the whole RNN again (from the last n steps). This would work for small values of `n` but due to the quadratic costs it will be too heavy for large `n`, like 500 or 1000.

**Comment by Trinkle23897** [comment]
> Or only its starting point is 0 hidden vector and the critic still uses the stack_num amount of state updates?

This one

> If I understand the proposal correctly this would mean that at each step t we compute the whole RNN again (from the last n steps). This would work for small values of n but due to the quadratic costs it will be too heavy for large n, like 500 or 1000.

Yes but actually RNN is not good for the long-horizon setting. Also you need to make sure the sampled data for training is correlated somehow. Otherwise you still need to recompute all `stack_num` states, e.g., two data points: `t[0:500]` and `t[2500:3000]` (1000 states), versus `t[0:500]` and `t[1:501]` (501 states)
